shell/server/proxy.js

```diff
@@ -129,38 +130,68 @@ Meteor.methods({
     // running.
 
     check(grainId, String);
-
-    var sessionId = Random.id();
-    var user = Meteor.user();
-    var userId = user ? user._id : undefined;
-
-    // Start the grain if it is not running.
-    var runningGrain = runningGrains[grainId];
-    var grainInfo;
-    if (runningGrain) {
-      grainInfo = waitPromise(runningGrain);
-    } else {
-      grainInfo = continueGrain(grainId);
+    var grain = Grains.findOne(grainId);
+    if (!grain) {
+      throw new Meteor.Error(403, "Unauthorized", "No grain found.");
     }
+    var title = grain.title;
 
-    updateLastActive(grainId, userId);
+    if (!mayOpenGrain(grainId, this.userId)) {
+      throw new Meteor.Error(403, "Unauthorized", "User is not authorized to access this grain.");
+    }
 
-    var isOwner = grainInfo.owner === userId;
+    return openSessionInternal(grainId, Meteor.user(), title);
+  },
 
-    var proxy = new Proxy(grainId, grainInfo.owner, sessionId, null, isOwner, user, null, false,
-                          grainInfo.supervisor);
-    proxies[sessionId] = proxy;
-    proxiesByHostId[proxy.hostId] = proxy;
+  openSessionFromApiToken: function(token) {
+    // Given an API token, either opens a new WebSession to the underlying grain or returns a
+    // path to which the client should redirect in order to open such a session.
 
-    Sessions.insert({
-      _id: sessionId,
-      grainId: grainId,
-      hostId: proxy.hostId,
-      timestamp: new Date().getTime(),
-      userId: userId
-    });
+    check(token, String);
+    var hashedToken = Crypto.createHash("sha256").update(token).digest("base64");
+    var apiToken = ApiTokens.findOne(hashedToken);
+    if (!apiToken) {
+      throw new Meteor.Error(403, "Unauthorized", "Invalid token.");
+    }
+    var grain = Grains.findOne({_id: apiToken.grainId});
+    if (!grain) {
+      throw new Meteor.error(403, "No such grain.");
```

Comment:
This case should definitely either be 404 or the user-visible part of the error should be "Unauthorized". I favor 404, as stated above.